src/rvsdg/egglog_optimizer/subst.rs

```diff
@@ -42,12 +44,144 @@ fn subst_all_rule(btype: String) -> String {
     )
 }
 
+// rtjoa: We could parameterize on `above`'s type (currently, always Body) to
+// support substituting under a Function as well
+fn subst_beneath_rules() -> Vec<String> {
+    let mut res = vec!["
+        (relation can-subst-Expr-beneath (Body Expr Expr))
+        (relation can-subst-Operand-beneath (Body Operand Operand))
+        (relation can-subst-Body-beneath (Body Body Body))
+        (relation can-subst-VecVecOperand-beneath (Body VecVecOperand VecVecOperand))
+        (relation can-subst-VecOperand-beneath (Body VecOperand VecOperand))
+
+        ;; Base case 'do the substitution' rules
+        (rule ((can-subst-Operand-beneath above from to)
+               (= above     (Theta from inputs outputs)))
+              ((union above (Theta to   inputs outputs)))
+              :ruleset subst)
+        (rule ((can-subst-VecOperand-beneath above from to)
+               (= above     (Theta pred inputs from)))
+              ((union above (Theta pred inputs to)))
+              :ruleset subst)
+        (rule ((can-subst-Operand-beneath above pred-from pred-to)
+               (can-subst-VecOperand-beneath above outputs-from outputs-to)
+               (= above     (Theta pred-from inputs outputs-from)))
+              ((union above (Theta pred-from inputs outputs-to)))
+              :ruleset subst)
+        (rule ((can-subst-VecVecOperand-beneath above from to)
+               (= above     (Gamma pred inputs from)))
+              ((union above (Gamma pred inputs to)))
+              :ruleset subst)
+        (rule ((can-subst-VecOperand-beneath above from to)
+               (= above     (OperandGroup from)))
+              ((union above (OperandGroup to)))
+              :ruleset subst)
+
+        ;; Learn can-subst-Operand-beneath
+        (rule ((can-subst-Body-beneath above from to)
+               (= new-from (Node from)))
+              ((can-subst-Operand-beneath above new-from (Node to)))
+              :ruleset subst)
+        (rule ((can-subst-Body-beneath above from to)
+               (= new-from (Project i from)))
+              ((can-subst-Operand-beneath above new-from (Project i to)))
+              :ruleset subst)
+
+        ;; Learn can-subst-body-beneath
+        (rule ((can-subst-Expr-beneath above from to)
+               (= new-from (PureOp from)))
+              ((can-subst-Body-beneath above new-from (PureOp to)))
+              :ruleset subst)
+        ;; Propagates up same context (Gamma: pred & inputs, Theta: inputs)
+        ;; rtjoa: Is it sound to propagate up outputs if we renumber args?
```

Comment:
If...
- The inputs to a Gamma `g` are `((Arg 2) ... more stuff)`
- `g` contains some expression `(Sub (Arg 0) (Arg 0))` in its outputs
- In the *surrounding body*, we have a can-subst-beneath showing that `(Sub (Arg 2) (Arg 2))` can be replaced with `some-eclass`
- Then within `g`, we can replace `(Sub (Arg 0) (Arg 0))` with `(SubstVecAll some-eclass g.inputs)`

However, this would require a lot of further thought, so I just left this mostly "talk to me" comment.

Reply:
I see, got it